beacon-chain/verification/blob.go

```diff
@@ -0,0 +1,285 @@
+	ErrSlotAboveFinalized           = errors.Wrap(ErrBlobInvalid, "slot <= finalized checkpoint")
+	ErrValidProposerSignature       = errors.Wrap(ErrBlobInvalid, "proposer signature could not be verified")
+	ErrSidecarParentSeen            = errors.Wrap(ErrBlobInvalid, "parent root has not been seen")
+	ErrSidecarParentValid           = errors.Wrap(ErrBlobInvalid, "parent block is not valid")
+	ErrSidecarParentSlotLower       = errors.Wrap(ErrBlobInvalid, "blob slot <= parent slot")
+	ErrSidecarDescendsFromFinalized = errors.Wrap(ErrBlobInvalid, "blob parent is not descended from the finalized block")
+	ErrSidecarInclusionProven       = errors.Wrap(ErrBlobInvalid, "sidecar inclusion proof verification failure")
+	ErrSidecarBlobCommitmentProven  = errors.Wrap(ErrBlobInvalid, "sidecar kzg commitment proof verification failed")
+	ErrSidecarProposerExpected      = errors.Wrap(ErrBlobInvalid, "sidecar was not proposed by the expected proposer_index")
+)
+
+type BlobVerifier struct {
+	*sharedResources
+	results              *results
+	blob                 blocks.ROBlob
+	parent               state.BeaconState
+	verifyBlobCommitment roblobCommitmentVerifier
+}
+
+type roblobCommitmentVerifier func(blocks.ROBlob) error
+
+// VerifiedROBlob "upgrades" the wrapped ROBlob to a VerifiedROBlob.
+// If any of the verifications ran against the blob failed, or some required verifications
+// were not run, an error will be returned.
+func (bv *BlobVerifier) VerifiedROBlob() (blocks.VerifiedROBlob, error) {
+	if bv.results.allSatisfied() {
+		return blocks.NewVerifiedROBlob(bv.blob), nil
+	}
+	return blocks.VerifiedROBlob{}, bv.results.errors(ErrBlobInvalid)
+}
+
+func (bv *BlobVerifier) recordResult(req Requirement, err *error) {
+	if err == nil || *err == nil {
+		bv.results.record(req, nil)
+		return
+	}
+	bv.results.record(req, *err)
+}
+
+// BlobIndexInBounds represents the follow spec verification:
+// [REJECT] The sidecar's index is consistent with MAX_BLOBS_PER_BLOCK -- i.e. blob_sidecar.index < MAX_BLOBS_PER_BLOCK.
+func (bv *BlobVerifier) BlobIndexInBounds() (err error) {
+	defer bv.recordResult(RequireBlobIndexInBounds, &err)
+	if bv.blob.Index >= fieldparams.MaxBlobsPerBlock {
+		log.WithFields(logging.BlobFields(bv.blob)).Debug("Sidecar index > MAX_BLOBS_PER_BLOCK")
+		return ErrBlobIndexInBounds
+	}
+	return nil
+}
+
+// SlotBelowMaxDisparity represents the spec verification:
+// [IGNORE] The sidecar is not from a future slot (with a MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance)
+// -- i.e. validate that block_header.slot <= current_slot
+func (bv *BlobVerifier) SlotBelowMaxDisparity() (err error) {
+	defer bv.recordResult(RequireSlotBelowMaxDisparity, &err)
+	if bv.clock.CurrentSlot() == bv.blob.Slot() {
+		return nil
+	}
+	// subtract the max clock disparity from the start slot time
+	validAfter := bv.clock.SlotStart(bv.blob.Slot()).Add(-1 * params.BeaconNetworkConfig().MaximumGossipClockDisparity)
```

Comment:
good thing to flag. This code can't enqueue the block, so either:
- the gossip code can use this check, and based on the error return, decide to do the more tolerant check and then enqueue
- have gossip implement that logic itself and delete this verification 
- make this a noop

I lean towards the first option because I think we should allow gossip to get a `VerifiedROBlob` if this verification fails. It *can* however run verification after the block has sat in the queue the right amount of time.